### PR TITLE
Topology to svg

### PR DESCRIPTION
**Pull request by sirocchj**
This is still wip but the idea is to create an sbt plugin to achieve something similar to https://zz85.github.io/kafka-streams-viz/

---
DONE:
- [x] parse Kafka Stream's topology description
- [x] convert it to `.dot` (see [here](https://en.wikipedia.org/wiki/DOT_(graph_description_language)))
- [x] use JS wrapper of `graphviz` to convert `.dot` to `.svg`
- [x] use `roughjs` to make the drawing appear more hand drawn
---
TODO:
- [ ] convert it into a plugin (it's missing all the relevant code)
- [ ] make it scan classloader in search of topologies
- [ ] allow topologies to be provided by other means

### CODE REVIEW

**Comment by sirocchj** [comment]
this means going from
```
Topology
Sub-topologies:
Sub-topology: 0
	Source:  KSTREAM-SOURCE-0000000000 (topics: [conversation-meta])
	--> KSTREAM-TRANSFORM-0000000001
	Processor: KSTREAM-TRANSFORM-0000000001 (stores: [conversation-meta-state])
	--> KSTREAM-KEY-SELECT-0000000002
	<-- KSTREAM-SOURCE-0000000000
	Processor: KSTREAM-KEY-SELECT-0000000002 (stores: [])
	--> KSTREAM-FILTER-0000000005
	<-- KSTREAM-TRANSFORM-0000000001
	Processor: KSTREAM-FILTER-0000000005 (stores: [])
	--> KSTREAM-SINK-0000000004
	<-- KSTREAM-KEY-SELECT-0000000002
	Sink: KSTREAM-SINK-0000000004 (topic: count-resolved-repartition)
	<-- KSTREAM-FILTER-0000000005
Sub-topology: 1
	Source: KSTREAM-SOURCE-0000000006 (topics: [count-resolved-repartition])
	--> KSTREAM-AGGREGATE-0000000003
	Processor: KSTREAM-AGGREGATE-0000000003 (stores: [count-resolved])
	--> KTABLE-TOSTREAM-0000000007
	<-- KSTREAM-SOURCE-0000000006
	Processor: KTABLE-TOSTREAM-0000000007 (stores: [])
	--> KSTREAM-SINK-0000000008
	<-- KSTREAM-AGGREGATE-0000000003
	Sink: KSTREAM-SINK-0000000008 (topic: streams-count-resolved)
	<-- KTABLE-TOSTREAM-0000000007
```

to

```
digraph G {
	label = "Kafka Streams Topology"
	subgraph cluster_0 {
		label = " Sub-\ntopology: 0";
		style = filled;
		color = lightgrey;
		node [style = filled, color = white];
		"KSTREAM-\nSOURCE-\n0000000000" -> "KSTREAM-\nTRANSFORM-\n0000000001";
		"KSTREAM-\nTRANSFORM-\n0000000001" -> "KSTREAM-\nKEY-\nSELECT-\n0000000002";
		"KSTREAM-\nKEY-\nSELECT-\n0000000002" -> "KSTREAM-\nFILTER-\n0000000005";
		"KSTREAM-\nFILTER-\n0000000005" -> "KSTREAM-\nSINK-\n0000000004";
	}
	subgraph cluster_1 {
		label = " Sub-\ntopology: 1";
		style = filled;
		color = lightgrey;
		node [style = filled, color = white];
		"KSTREAM-\nSOURCE-\n0000000006" -> "KSTREAM-\nAGGREGATE-\n0000000003";
		"KSTREAM-\nAGGREGATE-\n0000000003" -> "KTABLE-\nTOSTREAM-\n0000000007";
		"KTABLE-\nTOSTREAM-\n0000000007" -> "KSTREAM-\nSINK-\n0000000008";
	}
	"conversation-\nmeta" -> "KSTREAM-\nSOURCE-\n0000000000";
	"KSTREAM-\nTRANSFORM-\n0000000001" -> "conversation-\nmeta-\nstate";
	"KSTREAM-\nSINK-\n0000000004" -> "count-\nresolved-\nrepartition";
	"count-\nresolved-\nrepartition" -> "KSTREAM-\nSOURCE-\n0000000006";
	"KSTREAM-\nAGGREGATE-\n0000000003" -> "count-\nresolved";
	"KSTREAM-\nSINK-\n0000000008" -> "streams-\ncount-\nresolved";
	"conversation-\nmeta" [shape = rect];
	"count-\nresolved-\nrepartition" [shape = rect];
	"streams-\ncount-\nresolved" [shape = rect];
	"conversation-\nmeta-\nstate" [shape = cylinder];
	"count-\nresolved" [shape = cylinder];
}
```

and ultimately to

topology

File: src/main/scala/eisner/js/js.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  private[this] final def viz(dot: String): Future[String] =
+    engine.invokeFunction("viz", dot).asInstanceOf[CompletionStage[JMap[String, Object]]].toScala.flatMap {
+      case jm => Option(jm.get("result")) match {
+        case Some(s: String) => Future.successful(s)
+        case other => Future.failed(new RuntimeException(s"Could not find svg in viz.js response: $other"))
+      }
+    }
+
+  final def toSVG(dot: String): Future[Node] = viz(dot).flatMap(svg => Future.fromTry(Try(XML.loadString(svg))))
```
Comment (barambani) [inline src/main/scala/eisner/js/js.scala]: I guess you don't want to add a dependency to a proper effect system here, still ... Future ... 🤣

Comment (sirocchj) [reply src/main/scala/eisner/js/js.scala]: lol yeah, I’d like to minimise the dependencies but let’s see

**Comment by sirocchj** [comment]
@barambani I think I'll just merge this as is for now. I'm sure there's many things you won't like but let's address them after, wdyt?

**Pull request merged**